core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java

```diff
@@ -11,74 +11,253 @@
+                    PageStore pageStore, CacheEvictor evictor) {
+    mFsContext = fsContext;
+    mMetaStore = metaStore;
+    mPageStore = pageStore;
+    mEvictor = evictor;
+    mPageSize = (int) mFsContext.getClusterConf().getBytes(PropertyKey.USER_CLIENT_CACHE_PAGE_SIZE);
+    mCacheSize = mFsContext.getClusterConf().getBytes(PropertyKey.USER_CLIENT_CACHE_SIZE);
+    mLockSize = mFsContext.getClusterConf().getInt(PropertyKey.USER_CLIENT_CACHE_LOCK_SIZE);
+    mPageLocks = new ReentrantReadWriteLock[mLockSize];
+    for (int i = 0; i < mLockSize; i++) {
+      mPageLocks[i] = new ReentrantReadWriteLock();
+    }
+  }
+
+  /**
+   * Gets the lock for a particular page. Note that multiple pages may share the same lock as lock
+   * striping is used to reduce resource overhead for locks.
+   *
+   * @param fileId file identifier
+   * @param pageIndex index of the page within the file
+   * @return the corresponding page lock
+   */
+  private ReadWriteLock getPageLock(long fileId, long pageIndex) {
+    return mPageLocks[(int) (fileId + pageIndex) % mLockSize];
+  }
+
+  /**
+   * Gets a pair of locks to operate two given pages. One MUST acquire the first lock followed by
+   * the second lock.
+   *
+   * @param fileId file identifier
+   * @param pageIndex index of the page within the file
+   * @param fileId2 file identifier
+   * @param pageIndex2 index of the page within the file
+   * @return the corresponding page lock pair
+   */
+  private Pair<ReadWriteLock, ReadWriteLock> getPageLockPair(long fileId, long pageIndex,
+      long fileId2, long pageIndex2) {
+    if (fileId + pageIndex < fileId2 + pageIndex2) {
+      return new Pair<>(getPageLock(fileId, pageIndex), getPageLock(fileId2, pageIndex2));
+    } else {
+      return new Pair<>(getPageLock(fileId2, pageIndex2), getPageLock(fileId, pageIndex));
+    }
   }
 
   /**
-   * Writes a new page from a source channel to the store.
+   * Writes a new page from a source channel with best effort.
    *
-   * @param pageId page ID
+   * @param fileId file identifier
+   * @param pageIndex index of the page within the file
    * @param src source channel to read this new page
    * @throws IOException
    * @return the number of bytes written
    */
-  int put(long pageId, ReadableByteChannel src) throws IOException {
-    mMetaStore.addPage(pageId);
-    mPageStore.put(pageId, src);
-    return 0;
+  public int put(long fileId, long pageIndex, ReadableByteChannel src) throws IOException {
+    long victimFileId = 0;
+    long victimPageIndex = 0;
+
+    ReadWriteLock pageLock = getPageLock(fileId, pageIndex);
+    try (LockResource r = new LockResource(pageLock.writeLock())) {
+      boolean alreadyCached;
+      boolean needEvict = false;
+      try (LockResource r2 = new LockResource(mMetaLock.writeLock())) {
+        alreadyCached = mMetaStore.hasPage(fileId, pageIndex);
+        if (!alreadyCached) {
+          needEvict = (mPageSize + mMetaStore.size()) > mCacheSize;
```

Comment:
shouldn't it be `mPageStore.size()` based on my convo with @ZacBlanco

Reply:
I don't see it on the `PageStore` interface yet. Should I go ahead and add it?

Reply:
Refactored the method to `PageStore`.